packages/muelu/src/Graph/UncoupledAggregation/MueLu_UncoupledAggregationFactory_kokkos_def.hpp

```diff
@@ -284,79 +304,109 @@ namespace MueLu {
     }
     else
     {
-      SubFactoryMonitor sfm(*this, "Algo \"Graph Coloring\"", currentLevel);
+      
+      // We have the option to re-use coloring data between iterations of the factory
+      const std::string coloring_data_name = "reuse_coloring_data";
+      bool build_coloring = true;
+      const bool reuse_coloring = pL.get<bool>("aggregation: reuse coloring");
+
+      // Existing coloring data is stored as user-data in the currentLevel
+      if(reuse_coloring and currentLevel.IsAvailable(coloring_data_name))
+        build_coloring = false;
+        
+      TEUCHOS_TEST_FOR_EXCEPTION(reuse_coloring && (not pL.get<bool>("aggregation: deterministic")), 
+                                 Exceptions::RuntimeError, 
+                                 "MueLu::UncoupledAggregationFactory::Build: Request to reuse coloring requires aggregation to be deterministic!");
```

Comment:
Why do we need deterministic aggregation for color reuse?

Reply:
Yup, it appears that it will change the coloring each iteration, so we need it to be deterministic. I assume. I get segfaults if I allow coloring reuse with non-deterministic.

Reply:
I don't think so. We just need valid D2 coloring. The non-deterministic one should also work unless I'm missing something.

Reply:
Still tracking this down. Currently this error appears to be tied to the reuse of the level 1 aggregation algorithm `AggregationPhase1Algorithm_kokkos::BuildAggregatesRandom` (and likely others), which results in a `numNonAggragatedNodes<0` condition.

Reply:
Wait, numNonAggregatedNodes is actually negative?

Reply:
I see. Is there an easy way to check if the graph has changed (without storing the previous graph)?

Reply:
Well, we are calling the aggregation because we want to generate a new graph. So I think the assumption should be that it changes.